File: xenon-android-library/src/main/kotlin/com/abubusoft/xenon/context/XenonBeanContext.java
package com.abubusoft.xenon.context;

import java.lang.reflect.Field;

import com.abubusoft.kripton.android.Logger;

import android.content.Context;

/**
 * <p>
 * Utilità per la creazione di componenti argon.
 * 
 * @author Francesco Benincasa
 * 
 */
public abstract class XenonBeanContext {

	/**
	 * <p>
	 * Crea un'instanza di un determinato oggetto. E' richiesto che tale oggetto abbia il costruttore di default. Tutti i field con annotazione XenonBeanInject vengono risolti con gli oggetti già instanziati.
	 * </p>
	 * 
	 * @param clazz
	 * 		classe da istanziare
	 * @return
	 * 		crea una nuova istanza
	 */
	public static <E> E createInstance(Class<E> clazz) {
		E object = null;
		try {
			// creiamo istanza oggetto
			Logger.debug("BeanInject - createInstance of allocation %s", clazz.getSimpleName());
			object = clazz.newInstance();
		} catch (Exception e) {
			e.printStackTrace();
			Logger.error(e.getMessage());
			throw (new RuntimeException(e));
		}

		fillBean(object);

		return object;
	}

	/**
	 * <p>
	 * Effettua nuovamente il binding delle proprietà se eventualmente ci sono.
	 * </p>
	 * 
	 * @param bean
	 */
	public static void fillBean(Object bean) {
		// verifichiamo se è un argonComponent
		// if (clazz.getAnnotation(XenonBean.class) != null) {
		Field[] fields = bean.getClass().getFields();
		XenonBeanInject inject;
		int n = fields.length;
		for (int i = 0; i < n; i++) {
			inject = fields[i].getAnnotation(XenonBeanInject.class);

			if (inject != null) {
				try {
					fields[i].set(bean, inject.value().value);
					Logger.debug("BeanInject - Injected field %s with bean %s", fields[i].getName(), inject.value().toString());
				} catch (IllegalAccessException | IllegalArgumentException e) {
					e.printStackTrace();
					Logger.error(e.getMessage());
				}

			}
		}
		// }
	}

	/**
	 * <p>
	 * Imposta un bean predefinito.
	 * </p>
	 * @param type
	 * 		tipo da impostare
	 * @param value
	 * 		istanza
	 * 		
	 */
	public static void setBean(XenonBeanType type, Object value) {
		if (value != null)
			Logger.debug("BeanInject - setBean %s of allocation %s", type.toString(), value.getClass().getName());
		else {
			Logger.debug("BeanInject - setBean %s as null", type.toString());
		}
		type.value = value;
	}

	/**
	 * <p>
	 * Recupera un bean predefinito.
	 * </p>
	 * 
	 * @param type
	 * @return bean predefinito
	 */
	@SuppressWarnings("unchecked")
	public static <T> T getBean(XenonBeanType type) {
		return (T) type.value;
	}

	/**
	 * <p>
	 * Permette di recuperare da qualunque punto il context
	 * </p>
	 * 
	 * @return
	 * 		contesto android
	 */
	public static Context getContext() {
		return getBean(XenonBeanType.CONTEXT);
	}
}

File: xenon-android-library/src/main/kotlin/com/abubusoft/xenon/context/XenonBeanContext.kt
package com.abubusoft.xenon.context

import android.content.Context
import com.abubusoft.kripton.android.Logger

/**
 *
 *
 * Utilità per la creazione di componenti argon.
 *
 * @author Francesco Benincasa
 */
object XenonBeanContext {
    /**
     *
     *
     * Crea un'instanza di un determinato oggetto. E' richiesto che tale oggetto abbia il costruttore di default. Tutti i field con annotazione XenonBeanInject vengono risolti con gli oggetti già instanziati.
     *
     *
     * @param clazz
     * classe da istanziare
     * @return
     * crea una nuova istanza
     */
    fun <E> createInstance(clazz: Class<E>): E? {
        var `object`: E? = null
        `object` = try {
            // creiamo istanza oggetto
            Logger.debug("BeanInject - createInstance of allocation %s", clazz.simpleName)
            clazz.newInstance()
        } catch (e: Exception) {
            e.printStackTrace()
            Logger.error(e.message)
            throw RuntimeException(e)
        }
        fillBean(`object`)
        return `object`
    }

    /**
     *
     *
     * Effettua nuovamente il binding delle proprietà se eventualmente ci sono.
     *
     *
     * @param bean
     */
    fun fillBean(bean: Any?) {
        // verifichiamo se è un argonComponent
        // if (clazz.getAnnotation(XenonBean.class) != null) {
        val fields = bean!!.javaClass.fields
        var inject: XenonBeanInject?
        val n = fields.size
        for (i in 0 until n) {
            inject = fields[i].getAnnotation(XenonBeanInject::class.java)
            if (inject != null) {
                try {
                    fields[i][bean] = inject.value().value
                    Logger.debug("BeanInject - Injected field %s with bean %s", fields[i].name, inject.value().toString())
                } catch (e: IllegalAccessException) {
                    e.printStackTrace()
                    Logger.error(e.message)
                } catch (e: IllegalArgumentException) {
                    e.printStackTrace()
                    Logger.error(e.message)
                }
            }
        }
        // }
    }

    /**
     *
     *
     * Imposta un bean predefinito.
     *
     * @param type
     * tipo da impostare
     * @param value
     * istanza
     */
    fun setBean(type: XenonBeanType, value: Any?) {
        if (value != null) Logger.debug("BeanInject - setBean %s of allocation %s", type.toString(), value.javaClass.name) else {
            Logger.debug("BeanInject - setBean %s as null", type.toString())
        }
        type.value = value
    }

    /**
     *
     *
     * Recupera un bean predefinito.
     *
     *
     * @param type
     * @return bean predefinito
     */
    fun <T> getBean(type: XenonBeanType): T? {
        return type.value as T
    }

    /**
     *
     *
     * Permette di recuperare da qualunque punto il context
     *
     *
     * @return
     * contesto android
     */
    val context: Context
        get() = getBean<Context>(XenonBeanType.CONTEXT)!!
}